File: core/src/main/java/io/github/fourlastor/game/demo/round/step/StepState.java
package io.github.fourlastor.game.demo.round.step;

import com.badlogic.gdx.ai.msg.Telegram;
import io.github.fourlastor.game.demo.round.AbilityState;
import io.github.fourlastor.game.demo.state.GameState;
import java.util.function.Consumer;
import javax.inject.Inject;

public class StepState<T> extends AbilityState {

    private final Step<T> step;
    private final Consumer<T> next;
    private final Runnable cancel;

    public StepState(Step<T> step, Consumer<T> next, Runnable cancel) {
        this.step = step;
        this.next = next;
        this.cancel = cancel;
    }

    @Override
    public void enter(GameState state) {
        step.enter(state, next, cancel);
    }

    @Override
    public void update(GameState state) {
        step.update(state, next, cancel);
    }

    @Override
    public void exit(GameState state) {
        step.exit(state);
    }

    @Override
    public boolean onMessage(GameState state, Telegram telegram) {
        return step.onMessage(state, telegram);
    }

    public static class Factory {

        @Inject
        public Factory() {}

        public <T> StepState<T> create(Step<T> step, Consumer<T> next, Runnable cancel) {
            return new StepState<>(step, next, cancel);
        }
    }
}

File: core/src/main/java/io/github/fourlastor/game/demo/round/step/StepState.kt
package io.github.fourlastor.game.demo.round.step

import com.badlogic.gdx.ai.msg.Telegram
import io.github.fourlastor.game.demo.round.AbilityState
import io.github.fourlastor.game.demo.state.GameState
import java.util.function.Consumer
import javax.inject.Inject

class StepState<T>(private val step: Step<T>, private val next: Consumer<T>, private val cancel: Runnable) :
    AbilityState() {
    override fun enter(state: GameState) {
        step.enter(state, next, cancel)
    }

    override fun update(state: GameState) {
        step.update(state, next, cancel)
    }

    override fun exit(state: GameState) {
        step.exit()
    }

    override fun onMessage(state: GameState, telegram: Telegram): Boolean {
        return step.onMessage(state, telegram)
    }

    class Factory @Inject constructor() {
        fun <T> create(step: Step<T>, next: Consumer<T>, cancel: Runnable): StepState<T> =
            StepState(step, next, cancel)
    }
}
